wordpress-mobile/WordPress-iOS
# Remove entities with nil blog property

Refs #12028

Dependency: https://github.com/wordpress-mobile/WordPress-iOS-Shared/pull/227

As the main issue mentions, we're having a lot of crashes due to entities with `nil` blog property. We weren't able to determine WHY this is happening, but we were able to reproduce it by changing the `WordPress.sqlite` database.

Our attempt here is to sanitize Posts, Pages, Media, and Categories without a Blog.

Before diving into details, make sure you have https://sqlitebrowser.org/ installed on your computer.

## Testing

1. Run the app and detect where the `WordPress.sqlite` file is located (you can put a breakpoint in `ContextManager.m` or use [simsim](https://github.com/dsmelov/simsim).
2. Close the app
3. Open the database in SQLite Browser and run this query: `UPDATE ZBASEPOST SET ZBLOG = NULL` — now, all of your Posts and Pages will have `nil` Blog property
4. Close SQLite Browser and save the changes;
5. Open the app again
6. Check the `WordPress.sqlite` database again. The posts/pages without `Blog` were removed. Maybe you can have some entities without a `ZBLOG` — those are `ReaderPost` and they're not going to be deleted.
7. Load some posts, check some notifications: the app should work as it always worked. :)

## Additional info

This sanitization will happen once and every time the user updates the app. We added logging to keep track of how many users are affected by those orphaned entities.

## Reproducing the crash

Use `develop` for this.

1. First, look at your Notifications tab and find a comment notification. Take note on what Blog this post belong to;
2. Fresh install the app and login into your account;
3. Load the posts of the blog you received the notification in step 1 (do not open the Notifications tab)
4. Close the app;
5. Using the SQLite Browser, run this query: `UPDATE ZBASEPOST SET ZBLOG = NULL`
6. Open the app again, go to notifications. Tap on the notification and then to load the comments (you [can see where here](https://cld.wthms.co/LC0CE0)).
7. [The app will crash](https://cld.wthms.co/svrev5)

PR submission checklist:
